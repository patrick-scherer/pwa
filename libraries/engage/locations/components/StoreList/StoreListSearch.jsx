import React, {
  Fragment,
  memo,
  useCallback,
  useContext,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from 'react';
import PropTypes from 'prop-types';
import { i18n } from '@shopgate/engage/core';
import {
  Grid,
  InfoIcon,
  LocatorIcon,
  MagnifierIcon,
  MessageBar,
  ProgressBar,
} from '@shopgate/engage/components';
import { useCountriesNames } from '@shopgate/engage/i18n';
import { FulfillmentContext } from '../../locations.context';
import connect from './StoreListSearch.connector';
import {
  country,
  container,
  inputIcon,
  iconClass,
  input,
  messageClass,
  progressBar,
  queryLine,
  select,
  selectContainer,
} from './StoreListSearch.style';

/**
 * @param {Function} getProductLocations getProductLocations.
 * @param {Function} storeSearch .
 * @param {Object} search .
 * @returns {JSX}
 */
function StoreListSearch({
  getProductLocations,
  postalCode,
  countryCode,
  setPostalCode,
  setCountryCode,
}) {
  const {
    product,
    locations,
    shopSettings: { supportedCountries } = {},
  } = useContext(FulfillmentContext);

  const [loading, setLoading] = useState(false);
  const [message, setMessage] = useState('');
  const [inputPostalCode, setInputPostalCode] = useState(postalCode || '');

  const inputEl = useRef(null);

  useLayoutEffect(() => {
    if (!locations || locations.length === 0) {
      // Set a message when a location search resulted in zero locations.
      setMessage('locations.error_no_store_found');
    }
  }, [locations]);

  /**
   * Triggers a location update.
   * @param {Object} searchParams Parameters for the getProductLocation action call.
   * @param {string} [silent=false] If set to TRUE users will not get see any feedback popups when
   * geolocation tracking fails.
   */
  const updateProductLocations = useCallback(async (searchParams, silent = false) => {
    // Clear old error messages.
    setMessage('');
    setLoading(true);

    // Request new locations.
    const error = await getProductLocations(product.id, searchParams, silent);

    if (error) {
      setMessage(error);
    }

    setLoading(false);
  }, [getProductLocations, product.id]);

  /**
   * Triggers a product locations update at the initial rendering of the component. Potential
   * feedback popups on declined geolocation permissions will be suppressed.
   */
  useEffect(() => {
    let searchParams = null;

    if (postalCode !== null) {
      searchParams = {
        postalCode,
        countryCode,
      };
    }

    updateProductLocations(searchParams, true);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  /**
   * Triggers an update of Redux postalCode and countryCode and triggers fetching of new locations.
   * @param {string|null} postalCode Postal code for the update.
   * @param {string} [countryCode=countryCode] Country code for the update.
   */
  const update = useCallback((postalCodeUpdate, countryCodeUpdate = countryCode) => {
    if (postalCodeUpdate !== postalCode) {
      setPostalCode(postalCodeUpdate);
    }

    if (countryCodeUpdate !== countryCode) {
      setCountryCode(countryCodeUpdate);

      if (!postalCode) {
        // Prevent location update when the country selection changed, but the postal code is empty.
        return;
      }
    }

    let searchParams = null;

    if (postalCodeUpdate !== null) {
      searchParams = {
        postalCode: postalCodeUpdate,
        countryCode: countryCodeUpdate,
      };
    }

    updateProductLocations(searchParams);
  }, [countryCode, postalCode, setCountryCode, setPostalCode, updateProductLocations]);

  /**
   * Triggers an update when the value of the country selector changed.
   * @param {SyntheticEvent} event A React event object.
   */
  const handleCountrySelectChange = useCallback((event) => {
    update(postalCode, event.target.value);
  }, [update, postalCode]);

  /**
   * Blurs the postal code input to trigger an update.
   * @param {SyntheticEvent} event A React event object.
   */
  const handlePostalCodeSubmitKeyDown = useCallback((event) => {
    if (event.keyCode === 13) {
      inputEl.current.blur();
    }
  }, []);

  /**
   * Triggers an update when the input blurs.
   */
  const handlePostalCodeBlur = useCallback(() => {
    update(inputPostalCode);
  }, [inputPostalCode, update]);

  /**
   * Triggers an update when the locate me button was pressed. Also clears the local state for the
   * postal code input.
   */
  const handleLocateMeButton = useCallback(() => {
    setInputPostalCode('');
    update(null);
  }, [update]);

  /**
   * Updates the local state for the postal code input.
   * @param {SyntheticEvent} event A React event object
   */
  const handlePostalCodeChange = (event) => {
    setInputPostalCode(event.target.value);
  };

  const countries = useCountriesNames(supportedCountries);

  return (
    <Fragment>
      <div className={container}>
        <Grid>
          {supportedCountries && supportedCountries.length > 1 &&
            <Grid.Item shrink={0} className={country.toString()}>
              <div className={selectContainer}>
                <select
                  name="countryCode"
                  value={countryCode}
                  onChange={handleCountrySelectChange}
                  className={select}
                >
                  {
                    Object.keys(countries).map(key => (
                      <option className="option" value={key} key={key}>{countries[key]}</option>
                    ))
                  }
                </select>
              </div>
            </Grid.Item>
          }
          <Grid.Item grow={1} className={queryLine.toString()}>
            <span className={inputIcon} aria-hidden>
              <MagnifierIcon />
            </span>
            <input
              ref={inputEl}
              name="postalCode"
              className={input}
              value={inputPostalCode}
              onChange={handlePostalCodeChange}
              onBlur={handlePostalCodeBlur}
              onKeyDown={handlePostalCodeSubmitKeyDown}
              disabled={loading}
              type="search"
              autoComplete="off"
              autoCorrect="off"
              placeholder={i18n.text('locations.search_placeholder')}
            />
            <button
              onClick={handleLocateMeButton}
              type="button"
              className={inputIcon}
            >
              <LocatorIcon />
            </button>
          </Grid.Item>
        </Grid>
      </div>
      <div className={progressBar}>
        <ProgressBar isVisible={loading} />
      </div>
      {message &&
        <MessageBar
          messages={[{
            type: 'error',
            message,
            icon: InfoIcon,
          }]}
          classNames={{
            icon: iconClass,
            message: messageClass,
          }}
        />
      }
    </Fragment>
  );
}

StoreListSearch.defaultProps = {
  postalCode: null,
};

StoreListSearch.propTypes = {
  countryCode: PropTypes.string.isRequired,
  getProductLocations: PropTypes.func.isRequired,
  setCountryCode: PropTypes.func.isRequired,
  setPostalCode: PropTypes.func.isRequired,
  postalCode: PropTypes.string,
};

export default connect(memo(StoreListSearch));
