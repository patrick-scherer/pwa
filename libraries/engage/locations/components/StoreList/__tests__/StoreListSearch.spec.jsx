import React, { useContext } from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import StoreListSearch from '../StoreListSearch';

jest.mock('@shopgate/engage/components', () => ({
  LocatorIcon: () => '',
  MagnifierIcon: () => '',
  ProgressBar: () => '',
}));

jest.mock('react', () => ({
  ...require.requireActual('react'),
  useContext: jest.fn(),
}));
jest.mock('../StoreListSearch.connector', () => cmp => cmp);

describe('<StoreListSearch />', () => {
  const getProductLocations = jest.fn();
  const storeSearchQuery = jest.fn();

  const productId = 'ABC123';
  const postalCode = 'ACME';
  const defaultProps = {
    getProductLocations,
    storeSearchQuery,
    searchQuery: '',
  };
  const context = {
    product: { id: productId },
    locations: [{ code: 'LOCCODE' }],
  };

  beforeEach(() => {
    useContext.mockReturnValue(context);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render as expected', () => {
    const wrapper = shallow((
      <StoreListSearch {...defaultProps} />
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ProgressBar').prop('isVisible')).toEqual(false);
    expect(wrapper.find('input').prop('disabled')).toEqual(false);
  });

  it('should call getProductLocations with a postal code after filling the input and pressing enter', async () => {
    const wrapper = mount((
      <StoreListSearch {...defaultProps} />
    ));

    await act(async () => {
      await wrapper.find('input').simulate('change', { target: { value: postalCode } });
      await wrapper.find('input').simulate('keyDown', { keyCode: 13 });
    });

    expect(wrapper).toMatchSnapshot();
    expect(getProductLocations).toHaveBeenCalledTimes(1);
    expect(getProductLocations).toHaveBeenCalledWith(productId, postalCode);
    expect(wrapper.find('input').prop('value')).toEqual(postalCode);
  });

  it('should call getProductLocations after clicking the locate button', async () => {
    const wrapper = mount((
      <StoreListSearch {...defaultProps} />
    ));

    await act(async () => {
      await wrapper.find('input').simulate('change', { target: { value: postalCode } });
      await wrapper.find('button').simulate('click');
      await wrapper.update();
    });

    expect(wrapper).toMatchSnapshot();
    expect(getProductLocations).toHaveBeenCalledTimes(1);
    expect(getProductLocations).toHaveBeenCalledWith(productId, null);
    expect(wrapper.find('input').prop('value')).toEqual('');
  });
});
