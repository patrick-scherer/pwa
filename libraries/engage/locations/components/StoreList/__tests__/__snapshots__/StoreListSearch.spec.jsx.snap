// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > locations > components > StoreListSearch should call getProductLocations with a postal code after filling the input and pressing enter 1`] = `
<StoreListSearch
  countryCode="DE"
  getProductLocations={
    [MockFunction] {
      "calls": Array [
        Array [
          "ABC123",
          null,
          true,
        ],
        Array [
          "ABC123",
          Object {
            "countryCode": "DE",
            "postalCode": "ACME",
          },
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  postalCode="ACME"
  setCountryCode={[MockFunction]}
  setPostalCode={
    [MockFunction] {
      "calls": Array [
        Array [
          "ACME",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <Grid>
      <Grid.Item
        className="css-cclcsc"
        shrink={0}
      >
        <div
          className={
            Object {
              "data-css-1hrnuet": "",
            }
          }
        >
          <select
            className={
              Object {
                "data-css-nyfw00": "",
              }
            }
            name="countryCode"
            onChange={[Function]}
            value="DE"
          >
            <option
              className="option"
              key="DE"
              value="DE"
            >
              Germany
            </option>
            <option
              className="option"
              key="US"
              value="US"
            >
              United States
            </option>
          </select>
        </div>
      </Grid.Item>
      <Grid.Item
        className="css-retj1m"
        grow={1}
      >
        <span
          aria-hidden={true}
          className={
            Object {
              "data-css-1lgdfwm": "",
            }
          }
        >
          <MagnifierIcon />
        </span>
        <input
          autoComplete="off"
          autoCorrect="off"
          className={
            Object {
              "data-css-2h0g54": "",
            }
          }
          disabled={true}
          name="postalCode"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="locations.search_placeholder"
          type="search"
          value="ACME"
        />
        <button
          className={
            Object {
              "data-css-1lgdfwm": "",
            }
          }
          onClick={[Function]}
          type="button"
        >
          <LocatorIcon />
        </button>
      </Grid.Item>
    </Grid>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={true}
    />
  </div>
</StoreListSearch>
`;

exports[`engage > locations > components > StoreListSearch should render as expected 1`] = `
<Fragment>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <Grid>
      <Grid.Item
        className="css-cclcsc"
        shrink={0}
      >
        <div
          className={
            Object {
              "data-css-1hrnuet": "",
            }
          }
        >
          <select
            className={
              Object {
                "data-css-nyfw00": "",
              }
            }
            name="countryCode"
            onChange={[Function]}
            value="DE"
          >
            <option
              className="option"
              key="DE"
              value="DE"
            >
              Germany
            </option>
            <option
              className="option"
              key="US"
              value="US"
            >
              United States
            </option>
          </select>
        </div>
      </Grid.Item>
      <Grid.Item
        className="css-retj1m"
        grow={1}
      >
        <span
          aria-hidden={true}
          className={
            Object {
              "data-css-1lgdfwm": "",
            }
          }
        >
          <MagnifierIcon />
        </span>
        <input
          autoComplete="off"
          autoCorrect="off"
          className={
            Object {
              "data-css-2h0g54": "",
            }
          }
          disabled={false}
          name="postalCode"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="locations.search_placeholder"
          type="search"
          value=""
        />
        <button
          className={
            Object {
              "data-css-1lgdfwm": "",
            }
          }
          onClick={[Function]}
          type="button"
        >
          <LocatorIcon />
        </button>
      </Grid.Item>
    </Grid>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={false}
    />
  </div>
</Fragment>
`;
